File: frontend/src/lib/passwordPolicy.ts
/** SRS §4.3 REQ-6 — keep in sync with backend/app/utils/password_policy.py */
const HAS_DIGIT = /\d/;
const HAS_SPECIAL = /[^A-Za-z0-9]/;

export function isPasswordComplex(password: string): boolean {
  return (
    password.length >= 8 &&
    HAS_DIGIT.test(password) &&
    HAS_SPECIAL.test(password)
  );
}

export const PASSWORD_COMPLEXITY_HINT =
  "At least 8 characters, including a number and a special character.";

export function passwordComplexityError(password: string): string | null {
  if (isPasswordComplex(password)) return null;
  const parts: string[] = [];
  if (password.length < 8) parts.push("at least 8 characters");
  if (!HAS_DIGIT.test(password)) parts.push("a number");
  if (!HAS_SPECIAL.test(password)) parts.push("a special character");
  return `Password must include ${parts.join(", ")}.`;
}

File: backend/app/utils/password_policy.py
"""Password complexity — SRS §4.3 REQ-6."""
import re

_MIN_LENGTH = 8
_HAS_DIGIT = re.compile(r"\d")
_HAS_SPECIAL = re.compile(r"[^A-Za-z0-9]")


def validate_password_complexity(password: str) -> str | None:
    """Return an error message if the password fails policy, else None."""
    errors: list[str] = []
    if len(password) < _MIN_LENGTH:
        errors.append("at least 8 characters")
    if not _HAS_DIGIT.search(password):
        errors.append("at least one number")
    if not _HAS_SPECIAL.search(password):
        errors.append("at least one special character")
    if errors:
        return "Password must include " + ", ".join(errors) + "."
    return None
